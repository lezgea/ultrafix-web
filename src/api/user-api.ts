import { createApi } from '@reduxjs/toolkit/query/react';
import axiosBaseQuery from '@utils/axiosBaseQuery';
import { IActivateUserResponse, IChangeRequest, IForgetRequest, ILoginRequest, IRegisterRequest, IUser, LoginResponse, RegisterResponse } from './types/user-types';


export const userApi = createApi({
    reducerPath: 'userApi',
    baseQuery: axiosBaseQuery,
    tagTypes: ['User'],
    endpoints: (builder) => ({
        registerUser: builder.mutation<RegisterResponse, IRegisterRequest>({
            query: (credentials) => ({
                url: '/users',
                method: 'POST',
                data: credentials,
            }),
        }),
        loginUser: builder.mutation<LoginResponse, ILoginRequest>({
            query: (credentials) => ({
                url: '/users/login',
                method: 'POST',
                data: credentials,
            }),
        }),
        forgotPassword: builder.mutation<null, IForgetRequest>({
            query: (credentials) => ({
                url: '/users/forget-password',
                method: 'POST',
                data: credentials,
            }),
        }),
        changePassword: builder.mutation<null, IChangeRequest>({
            query: (credentials) => ({
                url: `/users/change-password?token=${encodeURIComponent(credentials.token)}`,
                method: 'POST',
                data: { password: credentials.password },
            }),
        }),
        logoutUser: builder.mutation<string, void>({
            query: () => ({
                url: '/users/logout',
                method: 'GET',
            }),
        }),
        activateUser: builder.query<IActivateUserResponse, { token: string }>({
            query: ({ token }) => ({
                url: `/users/active?token=${encodeURIComponent(token)}`,
                method: 'GET',
            }),
        }),
        getUser: builder.query<IUser, void>({
            query: () => ({
                url: '/users',
                method: 'GET',
            }),
            providesTags: ['User'],
        }),
        updateUser: builder.mutation<IUser, { id: number | string; data: Partial<IUser> }>({
            query: ({ id, data }) => ({
                url: `/users/${id}`,
                method: 'PUT',
                data,
            }),
            invalidatesTags: ['User'],
        }),
        deleteUser: builder.mutation<void, { id: number | string }>({
            query: ({ id }) => ({
                url: `/users/${id}`,
                method: 'DELETE',
            }),
            invalidatesTags: ['User'],
        }),
    }),
});

export const {
    useRegisterUserMutation,
    useLoginUserMutation,
    useLogoutUserMutation,
    useActivateUserQuery,
    useGetUserQuery,
    useUpdateUserMutation,
    useDeleteUserMutation,
    useForgotPasswordMutation,
    useChangePasswordMutation,
} = userApi;
