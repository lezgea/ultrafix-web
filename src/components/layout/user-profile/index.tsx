"use client";

import { useGetUserQuery, useLogoutUserMutation } from '@api/user-api';
import { ConfirmationModal, UserProfileSkeleton } from '@components/shared';
import Divider from '@components/shared/divider';
import { Dropdown } from '@components/shared/dropdown';
import { useAuthenticate } from '@hooks/use-auth';
import { createSelector } from '@reduxjs/toolkit';
import { logout } from '@slices/user-slice';
import { RootState } from '@store/store';
import getImgFromBase64 from '@utils/base64toImg';
import { Turret_Road } from 'next/font/google';
import Image from 'next/image';
import Link from 'next/link';
import { useRouter } from 'next/navigation';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';


const DROPDOWN_MENU: { route: string; label: string }[] = [
    { route: '/profile', label: 'Profile' },
    { route: '/races', label: 'Races' },
];


export const UserProfile: React.FC = () => {
    const [logoutUser, { isLoading, isError, error }] = useLogoutUserMutation();
    const [askModal, setAskModal] = React.useState<boolean>(false);
    const dispatch = useDispatch();
    const router = useRouter();

    const loading = useAuthenticate();

    const { user, isAuthenticated, loading: isUserLoading } = useSelector((state: RootState) => state.user);

    const userImage = React.useMemo(
        () => (user?.profileImage ? getImgFromBase64(user.profileImage) : '/svg/user.svg'),
        [user?.profileImage]
    );


    const handleLogout = async () => {
        try {
            await logoutUser().unwrap();
            dispatch(logout());
            router.push('/');
        } catch (error) {
            console.error('Logout failed', error);
        }
    };

    const DropdownContent = (
        <div role="menu" aria-orientation="vertical" aria-labelledby="options-menu" className="w-40">
            {DROPDOWN_MENU.map((item, index) => (
                <Link
                    key={index}
                    href={item.route}
                    className="block px-3 py-2 text-sm text-gray-700 hover:bg-gray-100 hover:text-primary rounded-md transition-all duration-200 ease-in-out"
                    role="menuitem"
                >
                    {item.label}
                </Link>
            ))}
            <Divider />
            <button
                disabled={isLoading}
                onClick={() => setAskModal(true)}
                className="flex w-full text-sm text-medium text-center justify-center px-5 py-2 text-gray-500 transition-all bg-gray-100 rounded-lg hover:bg-primaryDark hover:text-white shadow-neutral-300 hover:shadow-lg hover:shadow-neutral-300 hover:-tranneutral-y-px focus:shadow-none"
            >
                Sign Out
            </button>
        </div>
    );

    if (loading || isUserLoading) {
        return <UserProfileSkeleton />;
    }

    if (!isAuthenticated) {
        return (
            <div className="flex space-x-3">
                <Link href="/sign-in">
                    <button type="button" className="inline-flex w-auto text-center items-center px-4 py-2 text-white transition-all bg-primary rounded-lg sm:w-auto hover:bg-primaryDark hover:text-white shadow-neutral-300 dark:shadow-neutral-700 hover:shadow-lg hover:shadow-neutral-300 hover:-tranneutral-y-px focus:shadow-none">
                        Sign In
                    </button>
                </Link>
                <Link href="/sign-up">
                    <button type="button" className="hidden md:inline-flex w-auto text-center items-center px-4 py-2 text-white transition-all bg-gray-800 rounded-lg sm:w-auto hover:bg-dark hover:text-white shadow-neutral-300 dark:shadow-neutral-700 hover:shadow-lg hover:shadow-neutral-300 hover:-tranneutral-y-px focus:shadow-none">
                        Sign Up
                    </button>
                </Link>
            </div>
        );
    }

    return (
        <Dropdown content={DropdownContent}>
            <div className="flex items-center cursor-pointer group select-none">
                <div className="hidden md:flex text-gray-600 font-regmed mr-3 group-hover:text-primary transition-all duration-200 ease-in-out">
                    {user?.fullName?.split(" ")[0]}
                </div>
                <div className="relative w-[40px] h-[40px] min-w-[40px] min-h-[40px] rounded-full overflow-hidden">
                    <Image
                        src={userImage}
                        alt="Avatar"
                        fill={true}
                        className="object-cover"
                        priority={true}
                    />
                </div>
            </div>

            <ConfirmationModal
                visible={askModal}
                onConfirm={handleLogout}
                onClose={() => setAskModal(false)}
            />
        </Dropdown>
    );
};
