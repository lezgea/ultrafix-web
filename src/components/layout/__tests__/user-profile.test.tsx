import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import { useRouter } from 'next/navigation';
import { UserProfile } from '../user-profile';
import { useGetUserQuery, useLogoutUserMutation } from '@api/user-api';

// Mocking necessary hooks and modules
jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
}));

jest.mock('next/navigation', () => ({
    useRouter: jest.fn(),
}));

jest.mock('@api/user-api', () => ({
    useGetUserQuery: jest.fn(),
    useLogoutUserMutation: jest.fn(),
}));

describe('UserProfile Component', () => {
    const mockRouterPush = jest.fn();
    const mockDispatch = jest.fn();
    const mockLogoutUser = jest.fn();

    beforeEach(() => {
        (useRouter as jest.Mock).mockReturnValue({ push: mockRouterPush });
        (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
        (useLogoutUserMutation as jest.Mock).mockReturnValue([mockLogoutUser, { isLoading: false, isError: false }]);
        (useGetUserQuery as jest.Mock).mockReturnValue({
            data: { name: 'John Doe' },
            error: null,
            isLoading: false,
        });
    });

    it('should render the skeleton when loading', () => {
        (useSelector as unknown as jest.Mock).mockReturnValue({
            user: null,
            isAuthenticated: false,
            loading: true,
        });

        const { asFragment } = render(<UserProfile />);

        expect(screen.getByTestId('user-profile-skeleton')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render user profile and dropdown when authenticated', () => {
        (useSelector as unknown as jest.Mock).mockReturnValue({
            user: { fullName: 'John Doe', profileImage: null },
            isAuthenticated: true,
            loading: false,
        });

        const { asFragment } = render(<UserProfile />);

        const userName = screen.getByText('John Doe');
        expect(userName).toBeInTheDocument();
        const dropdownTrigger = screen.getByRole('img', { name: 'Avatar' });
        expect(dropdownTrigger).toBeInTheDocument();
        fireEvent.click(dropdownTrigger);
        const signOutButton = screen.getByText('Sign Out');
        expect(signOutButton).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

});