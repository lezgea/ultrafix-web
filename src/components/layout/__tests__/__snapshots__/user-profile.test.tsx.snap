// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile Component should render the skeleton when loading 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-end h-full"
    data-testid="user-profile-skeleton"
  >
    <div
      class="flex items-center space-x-3"
    >
      <div
        class="w-[80px] h-[20px] bg-gray-200 rounded-full"
      />
       
      <div
        class="relative w-[40px] h-[40px] rounded-full bg-gray-200 overflow-hidden"
      />
       
    </div>
  </div>
</DocumentFragment>
`;
